MRESOURCES-311. Ensure reproducible order in bundle

adoroszlai opened the pull request:
## What changes were proposed in this pull request?

`remote-resources:bundle` creates output that may not be reproducible.

https://github.com/apache/maven-remote-resources-plugin/blob/7a40047813a7cfde0a1bfeedff66bdc76f7ea430/src/main/java/org/apache/maven/plugin/resources/remote/BundleRemoteResourcesMojo.java#L108-L122

`DirectoryScanner` uses `java.io.File.list()`, which does not guarantee order.

https://issues.apache.org/jira/browse/MRESOURCES-311

## How was this patch tested?

Built the plugin and used it in another project.  Changed `naturalOrder` to `reverseOrder`, and repeated.  Verified output is reversed between the two runs.

elharo reviewed (commented):
Test?

slawekjaranowski commented:
I moved issue to correct jira project - https://issues.apache.org/jira/browse/MRRESOURCES-150
please update commit message and PR title

adoroszlai commented:
> Test?

Added.

> I moved issue to correct jira project

Thanks.

> please update commit message and PR title

PR title updated.

Would you like me to squash commits and force-push?

elharo commented on src/test/java/org/apache/maven/plugin/resources/remote/RemoteResourcesMojoTest.java:
```diff
@@ -322,6 +326,15 @@ protected void buildResourceBundle(String id, String sourceEncoding, String[] re
         }
     }
 
+    private static void assertContainsAllInOrder(Iterable<String> items, String data) {
```
Can you use Hamcrest's containsExactly for this?

slawekjaranowski commented:
> Would you like me to squash commits and force-push?

I can squash ....

slawekjaranowski commented on pom.xml:
```diff
@@ -211,6 +211,12 @@ under the License.
+    <dependency>
+      <groupId>org.hamcrest</groupId>
+      <artifactId>hamcrest-library</artifactId>
+      <version>1.3</version>
```
There is a 3.0 version 

https://hamcrest.org/JavaHamcrest/distributables

elharo commented on src/test/java/org/apache/maven/plugin/resources/remote/RemoteResourcesMojoTest.java:
```diff
@@ -51,6 +52,8 @@
 import org.eclipse.aether.internal.impl.SimpleLocalRepositoryManagerFactory;
 import org.eclipse.aether.repository.LocalRepository;
 
+import static org.assertj.core.api.Assertions.assertThat;
```
These days we're using Hamcrest, not assertj

slawekjaranowski replied:
It is only test dependencies  .... I will not require 
even more JUnit 5 doesn't use it

elharo replied:
I'd like to settle on one thing. I don't care much which one, but let's not have a random mix of libraries doing the same thing.

elharo replied:
My impression is that across Maven and even more broadly across Java, but definitely within Maven hamcrest is far more common now. I've certainly seen it a lot more while working my way through all the plugins than I've seen assert4j, though they are both present. JUnit 4 already depends on Hamcrest, though that got removed in JUnit 5.

slawekjaranowski replied:
ok we have Hamcrest so I would like to merge

slawekjaranowski commented:
@adoroszlai thanks

adoroszlai commented:
Thanks @elharo, @slawekjaranowski for the review.
